src/coprocessor/trie/mod.rs

```diff
@@ -0,0 +1,734 @@
+pub struct NewCoprocessor<F: LurkField> {
+    _p: PhantomData<F>,
+}
+
+impl<F: LurkField> Coprocessor<F> for NewCoprocessor<F> {
+    fn eval_arity(&self) -> usize {
+        0
+    }
+
+    fn simple_evaluate(&self, s: &mut Store<F>, _args: &[Ptr<F>]) -> Ptr<F> {
+        let trie: Trie<'_, F, 8, 55> = Trie::new(&s.poseidon_cache, &mut s.inverse_poseidon_cache);
+
+        let root = trie.root;
+
+        // FIXME: Use a custom type.
+        s.intern_num(Num::Scalar(root))
+    }
+}
+
+impl<F: LurkField> CoCircuit<F> for NewCoprocessor<F> {}
+
+#[derive(Clone, Debug, Serialize, Default, Deserialize)]
+pub struct LookupCoprocessor<F: LurkField> {
+    _p: PhantomData<F>,
+}
+
+impl<F: LurkField> Coprocessor<F> for LookupCoprocessor<F> {
+    fn eval_arity(&self) -> usize {
+        2
+    }
+
+    fn simple_evaluate(&self, s: &mut Store<F>, args: &[Ptr<F>]) -> Ptr<F> {
+        let root_ptr = args[0];
+        let key_ptr = args[1];
+        let root_scalar = *s.get_expr_hash(&root_ptr).unwrap().value();
+        let key_scalar = *s.get_expr_hash(&key_ptr).unwrap().value();
+        let trie: Trie<'_, F, 8, 55> = Trie::new_with_root(s, root_scalar);
+
+        let found = trie.lookup_aux(key_scalar).unwrap();
+
+        s.intern_maybe_opaque_comm(found)
+    }
+}
+
+impl<F: LurkField> CoCircuit<F> for LookupCoprocessor<F> {}
+
+#[derive(Clone, Debug, Serialize, Default, Deserialize)]
+pub struct InsertCoprocessor<F: LurkField> {
+    _p: PhantomData<F>,
+}
+
+impl<F: LurkField> Coprocessor<F> for InsertCoprocessor<F> {
+    fn eval_arity(&self) -> usize {
+        3
+    }
+
+    fn simple_evaluate(&self, s: &mut Store<F>, args: &[Ptr<F>]) -> Ptr<F> {
+        let root_ptr = args[0];
+        let key_ptr = args[1];
+        let val_ptr = args[2];
+        let root_scalar = *s.get_expr_hash(&root_ptr).unwrap().value();
+        let key_scalar = *s.get_expr_hash(&key_ptr).unwrap().value();
+        let val_scalar = *s.get_expr_hash(&val_ptr).unwrap().value();
+        let mut trie: Trie<'_, F, 8, 55> = Trie::new_with_root(s, root_scalar);
+        trie.insert(key_scalar, val_scalar).unwrap();
+
+        let new_root = trie.root;
+        s.intern_num(Num::Scalar(new_root))
+    }
+}
+
+impl<F: LurkField> CoCircuit<F> for InsertCoprocessor<F> {}
+
+/// Add the `Trie`-associated functions to a `Lang` with standard bindings.
+// TODO: define standard patterns for such modularity.
+pub fn install<F: LurkField>(s: &mut Store<F>, lang: &mut Lang<F, TrieCoproc<F>>) {
+    lang.add_binding((".lurk.trie.new", NewCoprocessor::default().into()), s);
+    lang.add_binding(
+        (".lurk.trie.lookup", LookupCoprocessor::default().into()),
+        s,
+    );
+    lang.add_binding(
+        (".lurk.trie.insert", InsertCoprocessor::default().into()),
+        s,
+    );
+}
+
+//pub type ChildMap<F: LurkField, const ARITY: usize> = HashMap<FWrap<F>, [F; ARITY]>;
+pub type ChildMap<F, const ARITY: usize> = InversePoseidonCache<F>;
+
+/// A sparse Trie.
```

Comment:
It would be nice to have documentation explaining the underlying details of our Trie. Is it described somewhere? What specific flavor of Trie are we using?

Reply:
I'll write a couple sentences. I don't think the 'flavor' implemented here necessarily exists on a universal menu of possible variations. But it's pretty simple. Big-endian bits of the (field element) key are taken N at a time, where the underlying hash has arity 2^n. This forms a fixed-length path whose final element is either zero or some non-zero payload. If the payload is assumed to be a commitment, then zero can be used as a hash whose preimage is undisocverable, and hence empty. To optimize creation of an empty tree, we precompute the empty subtree at each row rather than trying to actually construct the tree with a vast number of redundant hashes.

Reply:
Added as module doc, mostly just this text.